src/util/AsioHelpers.h

```diff
@@ -112,53 +112,69 @@ auto runFunctionOnExecutor(Executor executor, Function function,
 /// on an awaitable object that doesn't do this on its own. It's always better
 /// to integrate cancellation checks right into the awaitable itself, but
 /// sometimes this doesn't work because it's part of a library or boost asio
-/// itself. Once `timerRunning` resolves to `false` (or the awaitable is
-/// finished, whatever happens first), the cancellation checks are stopped.
-/// This needs to be called on a strand or in a single-threaded environment,
-/// otherwise this may lead to race conditions, due to issues with boost asio.
+/// itself. Once the value for `cancelCallback` is set and called by the caller
+/// of this function (or the awaitable is finished, whatever happens first), the
+/// cancellation checks are stopped. This needs to be called on a strand or in a
+/// single-threaded environment, otherwise this will lead to race conditions,
+/// due to issues with boost asio.
 template <typename T>
 inline net::awaitable<T> interruptible(
     net::awaitable<T> awaitable, ad_utility::SharedCancellationHandle handle,
-    std::shared_ptr<std::atomic_flag> timerRunning =
-        std::make_shared<std::atomic_flag>(true),
+    std::promise<std::function<void()>> cancelCallback,
     ad_utility::source_location loc = ad_utility::source_location::current()) {
   using namespace net::experimental::awaitable_operators;
+  auto timer =
+      std::make_shared<net::steady_timer>(co_await net::this_coro::executor);
```

Comment:
Am I understanding it correctly, that the reason why we need the explicit promise (previously bool) is that we want to be able to manually cancel the timer, to not have this unnecessary timer loop if we are not interested in it anymore. Just so that I understand this.

Reply:
Yes, exactly.